File: packages/core-go/jws.go
package arky

import (
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"strings"
)

var b64u = base64.RawURLEncoding // base64url, no padding

// protectedHeader builds the JWS protected header JSON.
//
// The member order is fixed at alg, typ, [kid,] b64, crit and written by hand
// rather than via encoding/json, which sorts map keys alphabetically. The
// header is base64url-encoded verbatim into the signing input, so its byte
// order is part of the signature: sorted keys would produce a different
// protected segment and a signature the other stacks reject.
func protectedHeader(kid string) string {
	var sb strings.Builder
	sb.WriteString(`{"alg":"EdDSA","typ":"JWS"`)
	if kid != "" {
		sb.WriteString(`,"kid":`)
		writeString(kid, &sb)
	}
	sb.WriteString(`,"b64":false,"crit":["b64"]}`)
	return sb.String()
}

// signingInput is the RFC 7797 detached-payload signing input:
// ASCII(BASE64URL(protected)) || '.' || payload.
func signingInput(protectedB64 string, payload []byte) []byte {
	buf := make([]byte, 0, len(protectedB64)+1+len(payload))
	buf = append(buf, protectedB64...)
	buf = append(buf, '.')
	return append(buf, payload...)
}

// SignDetached signs canonical payload bytes, returning the compact detached
// JWS <protected>..<signature> (RFC 7797, b64:false), per ARKY-TIM-v1 section 6.
// Pass kid = "" to omit the key id.
func SignDetached(payload []byte, priv ed25519.PrivateKey, kid string) string {
	protectedB64 := b64u.EncodeToString([]byte(protectedHeader(kid)))
	sig := ed25519.Sign(priv, signingInput(protectedB64, payload))
	return protectedB64 + ".." + b64u.EncodeToString(sig)
}

// DecodeProtectedHeader decodes the protected header of a compact JWS.
func DecodeProtectedHeader(jws string) (Value, error) {
	part, _, _ := strings.Cut(jws, ".")
	if part == "" {
		return nil, errors.New("jws: empty protected header")
	}
	raw, err := b64u.DecodeString(part)
	if err != nil {
		return nil, err
	}
	return Parse(string(raw))
}

// VerifyDetached verifies a detached compact JWS over payload against an
// Ed25519 public key. It returns true only if the signature is valid, the
// payload segment is empty, and the header is well-formed (alg EdDSA,
// b64:false, crit contains b64).
//
// The header checks are defense in depth: the protected header is part of the
// signing input, so tampering with alg already breaks the signature. They are
// kept so an alg:none downgrade is refused explicitly rather than incidentally.
// This function never panics on malformed input.
func VerifyDetached(jws string, payload []byte, pub ed25519.PublicKey) bool {
	parts := strings.Split(jws, ".")
	if len(parts) != 3 || parts[1] != "" {
		return false
	}
	protectedB64, sigB64 := parts[0], parts[2]

	hdr, err := DecodeProtectedHeader(jws)
	if err != nil {
		return false
	}
	obj, ok := hdr.(*Object)
	if !ok {
		return false
	}
	if alg, _ := obj.Get("alg"); alg != "EdDSA" {
		return false
	}
	if b64flag, _ := obj.Get("b64"); b64flag != false {
		return false
	}
	crit, _ := obj.Get("crit")
	critArr, ok := crit.([]Value)
	if !ok {
		return false
	}
	hasB64 := false
	for _, c := range critArr {
		if c == "b64" {
			hasB64 = true
			break
		}
	}
	if !hasB64 {
		return false
	}

	sig, err := b64u.DecodeString(sigB64)
	if err != nil || len(sig) != ed25519.SignatureSize {
		return false
	}
	if len(pub) != ed25519.PublicKeySize {
		return false
	}
	return ed25519.Verify(pub, signingInput(protectedB64, payload), sig)
}

File: packages/core-py/src/arky_core/jws.py
"""Detached-payload JWS (compact) with Ed25519/EdDSA.

Per ARKY-TIM-v1 section 6 and RFC 7797 (``b64:false``): the compact form is
``<protected>..<signature>`` and the payload — the JCS canonical bytes — is
supplied separately rather than embedded.
"""

from __future__ import annotations

import base64

from . import ed25519
from .canonicalize import _write_string
from .jsonvalue import JsonObject, parse

__all__ = [
    "b64url_decode",
    "b64url_encode",
    "decode_protected_header",
    "sign_detached",
    "verify_detached",
]


def b64url_encode(data: bytes) -> str:
    """base64url without padding."""
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def b64url_decode(text: str) -> bytes:
    """Decode base64url without padding. Raises on invalid input."""
    padding = "=" * (-len(text) % 4)
    return base64.urlsafe_b64decode(text + padding)


def _protected_header(kid: str | None) -> str:
    """Build the JWS protected header JSON.

    Member order is fixed at ``alg, typ, [kid,] b64, crit`` and written by hand
    rather than via ``json.dumps``, which would also insert spaces after
    separators by default. The header is base64url-encoded verbatim into the
    signing input, so both its byte order and its exact spelling are part of the
    signature: any difference here yields a signature the other stacks reject.
    """
    parts = ['{"alg":"EdDSA","typ":"JWS"']
    if kid:
        out: list[str] = []
        _write_string(kid, out)
        parts.append(',"kid":' + "".join(out))
    parts.append(',"b64":false,"crit":["b64"]}')
    return "".join(parts)


def _signing_input(protected_b64: str, payload: bytes) -> bytes:
    """RFC 7797 signing input: ASCII(BASE64URL(protected)) || '.' || payload."""
    return protected_b64.encode("ascii") + b"." + payload


def sign_detached(payload: bytes, seed: bytes, kid: str | None = None) -> str:
    """Sign canonical ``payload`` bytes, returning ``<protected>..<signature>``."""
    protected_b64 = b64url_encode(_protected_header(kid).encode("utf-8"))
    signature = ed25519.sign(_signing_input(protected_b64, payload), seed)
    return f"{protected_b64}..{b64url_encode(signature)}"


def decode_protected_header(jws: str) -> object:
    """Decode the protected header of a compact JWS. Raises on malformed input."""
    protected_b64 = jws.split(".", 1)[0]
    if not protected_b64:
        raise ValueError("jws: empty protected header")
    return parse(b64url_decode(protected_b64).decode("utf-8"))


def verify_detached(jws: str, payload: bytes, public_key: bytes) -> bool:
    """Verify a detached compact JWS against an Ed25519 public key.

    Returns True only if the signature is valid, the payload segment is empty,
    and the header is well-formed (``alg`` EdDSA, ``b64`` false, ``crit``
    containing ``b64``).

    The header checks are defense in depth: the protected header is part of the
    signing input, so tampering with ``alg`` already breaks the signature. They
    are kept so an ``alg:none`` downgrade is refused explicitly rather than
    incidentally. Never raises, whatever the input.
    """
    parts = jws.split(".")
    if len(parts) != 3 or parts[1] != "":
        return False
    protected_b64, sig_b64 = parts[0], parts[2]

    try:
        header = decode_protected_header(jws)
    except Exception:
        return False
    if not isinstance(header, JsonObject):
        return False
    if header.get("alg") != "EdDSA":
        return False
    if header.get("b64") is not False:
        return False
    crit = header.get("crit")
    if not isinstance(crit, list) or "b64" not in crit:
        return False

    try:
        signature = b64url_decode(sig_b64)
    except Exception:
        return False
    return ed25519.verify(signature, _signing_input(protected_b64, payload), public_key)
